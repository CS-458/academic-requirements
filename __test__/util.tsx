import { DndProvider } from "react-dnd";
import { HTML5Backend } from "react-dnd-html5-backend";
import { QueryClient, QueryClientProvider } from "react-query";
import { testApiHandler } from "next-test-api-route-handler";
import { NextApiHandler } from "next";

export async function fetchReplace(initial: string): Promise<Response> {
  const res = { resolve: (_a: any) => {}, reject: (_e: any) => {} };
  const result: Promise<Response> = new Promise((resolve, reject) => {
    res.resolve = resolve;
    res.reject = reject;
  });
  const [url, params] = initial.split("?");
  await testApiHandler({
    handler: await import(`../pages/${url}`),
    url: initial,
    params: params.split("&").reduce((a, b) => {
      const [name, value] = b.split("=");
      return { ...a, [name]: value };
    }, {}),
    test: async ({ fetch }) => {
      const inner = await fetch();
      // console.error(initial, await inner.json());
      // res.reject(new Error("Failed"));
      res.resolve(inner);
    }
  });
  return await result;
}

/**
 *  Wrap Elements with QueryClient & DndProvider similar to _app.tsx
 * */
export function wrapper(children: JSX.Element | JSX.Element[]): JSX.Element {
  // @ts-expect-error Parameters are different
  window.fetch = fetchReplace;
  // <DndProvider backend={HTML5Backend}>{children}</DndProvider>
  return (
    <QueryClientProvider client={new QueryClient()}>
      {children}
    </QueryClientProvider>
  );
}
