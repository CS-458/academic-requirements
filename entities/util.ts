let id = 0;
export function uniqueId(): string {
  return `id_${id++}`;
}
