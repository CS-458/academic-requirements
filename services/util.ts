/// Fetch a request (similar to fetch), but throw an error if the request fails
export async function fetchApi<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, init);
  if (!res.ok) {
    throw new Error(`Response code ${res.status} for request to ${url}`);
  }
  return await res.json();
}
