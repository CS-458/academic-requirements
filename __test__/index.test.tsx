import "@testing-library/jest-dom";
import { screen } from "@testing-library/react";

import Home from "../pages/index";
import { setupUser, render } from "./util";

test("Verify Majors and Concentrations", async () => {
  const user = setupUser();
  const index = render(<Home />);
  expect(index.baseElement).toMatchSnapshot();

  const generateButton = screen.getByText(/Generate Schedule/i);
  expect(generateButton).toBeDisabled();

  await user.selectAutocomplete(/Major/i, /^Computer Science$/i);
  expect(index.baseElement).toMatchSnapshot();
  expect(generateButton).toBeDisabled();

  await user.selectAutocomplete(/Concentration/i, /Interdisciplinary/i);
  expect(index.baseElement).toMatchSnapshot();

  expect(
    screen.getByLabelText(/Use Suggested Four Year Plan/i)
  ).not.toBeChecked();

  expect(generateButton).not.toBeDisabled();
});

test("Verify Four Year Plan Not Show", async () => {
  const user = setupUser();
  const index = render(<Home />);
  expect(index.baseElement).toMatchSnapshot();

  const generateButton = screen.getByText(/Generate Schedule/i);
  expect(generateButton).toBeDisabled();

  await user.selectAutocomplete(/Major/i, /^Psychology$/i);
  expect(index.baseElement).toMatchSnapshot();
  expect(generateButton).toBeDisabled();

  await user.selectAutocomplete(/Concentration/i, /Self-Designed/i);
  expect(index.baseElement).toMatchSnapshot();

  expect(screen.queryByLabelText(/Use Suggested Four Year Plan/i)).toBeNull();

  expect(generateButton).not.toBeDisabled();
});

test("Verify adding completed course", async () => {
  const user = setupUser();
  const index = render(<Home />);
  expect(index.baseElement).toMatchSnapshot();

  const addButton = screen.getByText(/Add Course/i);
  expect(addButton).toBeDisabled();

  await user.selectAutocomplete(/Course Subject/i, /AEC/i);
  expect(index.baseElement).toMatchSnapshot();
  expect(addButton).toBeDisabled();

  await user.selectAutocomplete(/Course Number/i, /191/i);
  expect(index.baseElement).toMatchSnapshot();

  expect(addButton).not.toBeDisabled();
  await user.click(addButton);
  expect(screen.getByText(/AEC-191/i)).toBeInTheDocument();
});

test("Verify deleting completed course", async () => {
  const user = setupUser();
  const index = render(<Home />);
  expect(index.baseElement).toMatchSnapshot();

  const addButton = screen.getByText(/Add Course/i);
  expect(addButton).toBeDisabled();

  await user.selectAutocomplete(/Course Subject/i, /AEC/i);
  expect(index.baseElement).toMatchSnapshot();
  expect(addButton).toBeDisabled();

  await user.selectAutocomplete(/Course Number/i, /191/i);
  expect(index.baseElement).toMatchSnapshot();

  expect(addButton).not.toBeDisabled();
  await user.click(addButton);
  expect(screen.getByText(/AEC-191/i)).toBeInTheDocument();

  const deleteButton = screen.getByTestId("delete-icon");
  await user.click(deleteButton);
  expect(screen.queryByText(/AEC-191/i)).toBeNull();
});
