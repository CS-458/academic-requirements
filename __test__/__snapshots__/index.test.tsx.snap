// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify Majors and Concentrations 1`] = `
<body>
  <div>
    <div>
      <div
        class="App"
      >
        <div
          data-testid="content"
        >
          <header
            class="Four-Year-Plan"
          >
            <h1>
              Academic Planner
            </h1>
          </header>
          <div
            class="overlay"
            data-testid="popup"
            style="visibility: hidden; opacity: 0;"
          >
            <div
              class="popup"
            >
              <h2>
                Error
              </h2>
              <span
                class="close"
              >
                Ã—
              </span>
              <div
                class="content"
              />
            </div>
          </div>
          <div
            class="screen"
          >
            <div
              class="input-grid"
            >
              <div
                class="input-grid-dropdown"
                data-testid="MajorDropDown"
              >
                <div
                  class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1qqsdnr-MuiAutocomplete-root"
                >
                  <div
                    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                      data-shrink="false"
                      for=":r0:"
                      id=":r0:-label"
                    >
                      Major
                    </label>
                    <div
                      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                    >
                      <input
                        aria-autocomplete="list"
                        aria-expanded="false"
                        aria-invalid="false"
                        autocapitalize="none"
                        autocomplete="off"
                        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                        id=":r0:"
                        role="combobox"
                        spellcheck="false"
                        type="text"
                        value=""
                      />
                      <div
                        class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                      >
                        <button
                          aria-label="Open"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                          tabindex="-1"
                          title="Open"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                            data-testid="ArrowDropDownIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M7 10l5 5 5-5z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                      <fieldset
                        aria-hidden="true"
                        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                      >
                        <legend
                          class="css-yjsfm1"
                        >
                          <span>
                            Major
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="input-grid-dropdown"
              />
              <div
                class="input-grid-item"
              >
                <div
                  class="courseDropdowns"
                >
                  <div
                    class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1qqsdnr-MuiAutocomplete-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for=":r2:"
                        id=":r2:-label"
                      >
                        Course Subject
                      </label>
                      <div
                        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-autocomplete="list"
                          aria-expanded="false"
                          aria-invalid="false"
                          autocapitalize="none"
                          autocomplete="off"
                          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                          id=":r2:"
                          role="combobox"
                          spellcheck="false"
                          type="text"
                          value=""
                        />
                        <div
                          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                        >
                          <button
                            aria-label="Open"
                            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                            tabindex="-1"
                            title="Open"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                              data-testid="ArrowDropDownIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M7 10l5 5 5-5z"
                              />
                            </svg>
                            <span
                              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                            />
                          </button>
                        </div>
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-yjsfm1"
                          >
                            <span>
                              Course Subject
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1qqsdnr-MuiAutocomplete-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for=":r4:"
                        id=":r4:-label"
                      >
                        Course Number
                      </label>
                      <div
                        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-autocomplete="list"
                          aria-expanded="false"
                          aria-invalid="false"
                          autocapitalize="none"
                          autocomplete="off"
                          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                          id=":r4:"
                          role="combobox"
                          spellcheck="false"
                          type="text"
                          value=""
                        />
                        <div
                          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                        >
                          <button
                            aria-label="Open"
                            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                            tabindex="-1"
                            title="Open"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                              data-testid="ArrowDropDownIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M7 10l5 5 5-5z"
                              />
                            </svg>
                            <span
                              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                            />
                          </button>
                        </div>
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-yjsfm1"
                          >
                            <span>
                              Course Number
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                  </div>
                </div>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  Add Course
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <div
                class="input-grid-item"
              />
              <div
                class="input-grid-item"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  Generate Schedule
                </button>
                <br />
              </div>
              <div
                class="input-grid-item-courses"
              >
                <div
                  class="completedCourses"
                >
                  <h2>
                    Completed Courses
                  </h2>
                  <div
                    class="courseList"
                    style="grid-template-columns: repeat(0.9, 1fr);"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="overlay"
        data-testid="popup"
        style="visibility: hidden; opacity: 0;"
      >
        <div
          class="popup"
        >
          <h2>
            Error
          </h2>
          <span
            class="close"
          >
            Ã—
          </span>
          <div
            class="content"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
